{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Question 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('3R', '21378977', '21381970') \"tin\";\n",
      "Linear genomic distance is 5375\n",
      "search tally is 12\n"
     ]
    }
   ],
   "source": [
    "#Open file\n",
    "fs = open('/Users/cmdb/data/genomes/BDGP6.Ensembl.81.gtf', 'r')\n",
    "\n",
    "counter = 0\n",
    "#Store the data in a list\n",
    "p_genes = []\n",
    "for i, line in enumerate(fs):\n",
    "    #1. skip first 5 lines (header)\n",
    "    if i < 5:\n",
    "        continue\n",
    "    #2. split the line into a list\n",
    "    fields = line.split()\n",
    "    #3. chrm, start, end, gene_name, feature_type, gene_biotype\n",
    "    chrm = fields[0]\n",
    "    start = fields[3]\n",
    "    end = fields[4]\n",
    "    #4. ignore entries not on 3R\n",
    "    if chrm == '3R':\n",
    "        #5. ignore non \"gene\" entries\n",
    "        if fields[2] == 'gene':\n",
    "            feature_type = fields[2]\n",
    "            #6. ignore non-protein-encoding genes\n",
    "            if 'protein_coding' in fields[-1]:\n",
    "                #7. store gene_name and gene_biotype\n",
    "                gene_name = fields[13]\n",
    "                gene_biotype = fields[-1]\n",
    "                p_genes += [(chrm, start, end, feature_type, gene_name, gene_biotype)]\n",
    "                counter += 1\n",
    "                #if counter < 10:\n",
    "                    #print(chrm, start, end, feature_type, gene_name, gene_biotype)\n",
    "                    #print(type(start))\n",
    "                    \n",
    "                    \n",
    "#print('len of p_genes is', len(p_genes))\n",
    "\n",
    "#binary search through p_genes list:\n",
    "\n",
    "search_pos = 21378950\n",
    "#initialize low, high, and mid\n",
    "low = 0\n",
    "high = len(p_genes)-1\n",
    "mid = (len(p_genes))//2\n",
    "\n",
    "gene = ''\n",
    "\n",
    "search_tally = 0\n",
    "while True:\n",
    "    #print()\n",
    "    #print('low is index', low)\n",
    "    #print('high is index', high)\n",
    "    #print('mid is index', mid)\n",
    "    #print('search_pos is', search_pos)\n",
    "    #print('tally is', search_tally)\n",
    "    \n",
    "    #print('p_genes[mid][1] is', p_genes[mid][1])\n",
    "    #print('p_genes[mid][2] is', p_genes[mid][2])\n",
    "        \n",
    "    if high == low + 1:\n",
    "        #print('high == low + 1')\n",
    "        #find distance to end of low\n",
    "        distance_to_low = search_pos - int(p_genes[low][2])\n",
    "        #print('distance_to_low is', distance_to_low)\n",
    "        #find distance to start of high\n",
    "        distance_to_high = int(p_genes[high][1]) - search_pos\n",
    "        #print('distance_to_high is', distance_to_high)\n",
    "        \n",
    "        #compare, figure out which is closest (i.e., which has smaller distance)\n",
    "        if distance_to_low < distance_to_high:\n",
    "            #low is the gene!\n",
    "            #print('low is the gene')\n",
    "            #print('the gene is', p_genes[low][4])\n",
    "            print(p_genes[low][0:3], p_genes[low][4])\n",
    "            lin_dist = distance_to_low\n",
    "            \n",
    "        else:\n",
    "            #high is the gene!\n",
    "            #print('high is the gene')\n",
    "            #print(p_genes[high][4])\n",
    "            print(p_genes[high][0:3], p_genes[high][4])\n",
    "            lin_dist = distance_to_low\n",
    "        \n",
    "        break\n",
    "        \n",
    "    if search_pos < int(p_genes[mid][1]):\n",
    "        high = mid\n",
    "        #print('search_pos is less than mid')\n",
    "        #print('high is now mid:', mid)\n",
    "        mid = (high+low)//2\n",
    "        #print('mid is now', mid)\n",
    "    elif search_pos > int(p_genes[mid][2]):\n",
    "        low = mid\n",
    "        #print('search_pos is greater than mid')\n",
    "        #print('low is now mid:', low)\n",
    "        mid = (high+low)//2\n",
    "        #print('mid is now', mid)\n",
    "        \n",
    "    else:\n",
    "        #we found the gene!\n",
    "        #output gene_name\n",
    "        print('the gene is', p_genes[mid][4])\n",
    "        break\n",
    "    search_tally += 1\n",
    "    #print('~~~~~~~~~~~~~~~~')\n",
    "    \n",
    "print('Linear genomic distance is', lin_dist)\n",
    "print('search tally is', search_tally)\n",
    "        \n",
    "#Report the gene's linear genomic distance from 3R:21,378,950.\n",
    "#How many iterations (a tally), did it take the search procedure \n",
    "#to find the nearest gene? Report this.\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
